File: scripts/apply_db_schema.mjs
#!/usr/bin/env node
import fs from 'node:fs'
import path from 'node:path'
import process from 'node:process'
import mysql from 'mysql2/promise'

function loadDotEnv(filePath) {
  if (!fs.existsSync(filePath)) return

  for (const rawLine of fs.readFileSync(filePath, 'utf8').split(/\r?\n/)) {
    const line = rawLine.trim()
    if (!line || line.startsWith('#') || !line.includes('=')) continue

    const [rawKey, ...rawValueParts] = line.split('=')
    const key = rawKey.trim()
    const value = rawValueParts.join('=').trim().replace(/^['"]|['"]$/g, '')
    if (!process.env[key]) process.env[key] = value
  }
}

function stripSqlComments(sql) {
  return sql
    .split(/\r?\n/)
    .filter((line) => !line.trim().startsWith('--'))
    .join('\n')
}

function splitSqlStatements(sql) {
  return stripSqlComments(sql)
    .split(';')
    .map((statement) => statement.trim())
    .filter(Boolean)
}

const projectRoot = path.resolve(import.meta.dirname, '..')
loadDotEnv(path.join(projectRoot, '.env.local'))

const requiredEnv = ['DB_HOST', 'DB_USER', 'DB_PASSWORD', 'DB_NAME']
const missing = requiredEnv.filter((key) => !process.env[key])
if (missing.length) {
  console.error(`Missing required DB env vars: ${missing.join(', ')}`)
  process.exit(2)
}

const schemaPath = path.join(projectRoot, 'database', 'schema.sql')
const statements = splitSqlStatements(fs.readFileSync(schemaPath, 'utf8'))

const connection = await mysql.createConnection({
  host: process.env.DB_HOST,
  port: Number(process.env.DB_PORT || 3306),
  user: process.env.DB_USER,
  password: process.env.DB_PASSWORD,
  database: process.env.DB_NAME,
  charset: 'utf8mb4',
  connectTimeout: Number(process.env.DB_CONNECT_TIMEOUT || 10000),
  multipleStatements: false,
})

try {
  await connection.beginTransaction()
  for (const statement of statements) {
    await connection.execute(statement)
  }
  await connection.commit()
  console.log(
    `Applied ${statements.length} SQL statements to ${process.env.DB_HOST}:${process.env.DB_PORT || 3306}/${process.env.DB_NAME} as ${process.env.DB_USER}`,
  )
} catch (error) {
  await connection.rollback()
  console.error(error)
  process.exitCode = 1
} finally {
  await connection.end()
}

File: scripts/apply_db_schema.py
#!/usr/bin/env python3
"""Apply database/schema.sql to the configured MariaDB database.

Configuration is loaded from environment variables first, then .env.local.
Required variables: DB_HOST, DB_USER, DB_PASSWORD, DB_NAME.
Optional: DB_PORT, DB_CONNECT_TIMEOUT.
"""

from __future__ import annotations

import os
import re
import sys
from pathlib import Path


def load_dotenv(path: Path) -> None:
    if not path.exists():
        return
    for raw_line in path.read_text().splitlines():
        line = raw_line.strip()
        if not line or line.startswith('#') or '=' not in line:
            continue
        key, value = line.split('=', 1)
        key = key.strip()
        value = value.strip().strip('"').strip("'")
        os.environ.setdefault(key, value)


def split_sql(sql: str) -> list[str]:
    # This schema does not define routines/triggers, so semicolon splitting is enough.
    cleaned = re.sub(r'^\s*--.*$', '', sql, flags=re.MULTILINE)
    return [stmt.strip() for stmt in cleaned.split(';') if stmt.strip()]


def main() -> int:
    project_root = Path(__file__).resolve().parents[1]
    load_dotenv(project_root / '.env.local')

    missing = [name for name in ['DB_HOST', 'DB_USER', 'DB_PASSWORD', 'DB_NAME'] if not os.environ.get(name)]
    if missing:
        print(f"Missing required DB env vars: {', '.join(missing)}", file=sys.stderr)
        return 2

    try:
        import pymysql
    except ImportError:
        print('PyMySQL is required. Install with: python3 -m pip install PyMySQL', file=sys.stderr)
        return 3

    schema_path = project_root / 'database' / 'schema.sql'
    statements = split_sql(schema_path.read_text())

    conn = pymysql.connect(
        host=os.environ['DB_HOST'],
        port=int(os.environ.get('DB_PORT', '3306')),
        user=os.environ['DB_USER'],
        password=os.environ['DB_PASSWORD'],
        database=os.environ['DB_NAME'],
        charset='utf8mb4',
        autocommit=False,
        connect_timeout=int(os.environ.get('DB_CONNECT_TIMEOUT', '10')),
    )

    try:
        with conn.cursor() as cur:
            for statement in statements:
                cur.execute(statement)
        conn.commit()
    except Exception:
        conn.rollback()
        raise
    finally:
        conn.close()

    print(f"Applied {len(statements)} SQL statements to {os.environ['DB_HOST']}:{os.environ.get('DB_PORT', '3306')}/{os.environ['DB_NAME']} as {os.environ['DB_USER']}")
    return 0


if __name__ == '__main__':
    raise SystemExit(main())
